File: app/src/main/java/com/s95ammar/weeklyschedule/views/fragments/ScheduleNamerDialog.java
package com.s95ammar.weeklyschedule.views.fragments;

import android.app.AlertDialog;
import android.app.Dialog;
import android.content.Context;
import android.content.DialogInterface;
import android.graphics.drawable.ColorDrawable;
import android.os.Bundle;
import androidx.appcompat.app.AppCompatDialogFragment;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.EditText;
import android.widget.Toast;

import com.s95ammar.weeklyschedule.R;

import java.util.HashSet;

import static com.s95ammar.weeklyschedule.views.fragments.SchedulesListFragment.SchedulesListManager.KEY_INDEX;
import static com.s95ammar.weeklyschedule.views.fragments.SchedulesListFragment.SchedulesListManager.KEY_NAME;

public class ScheduleNamerDialog extends AppCompatDialogFragment {
    private EditText EditTextName;
    private ScheduleNamerListener mListener;
    private static final String ADD_TITLE = "New Schedule";
    private static final String RENAME_TITLE = "Rename Schedule";

    public interface Action {
        int ADD = -1;
        // REMAKE = any value other than -1 (value will be used as index)
    }


    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        if (context instanceof ScheduleNamerListener) {
            mListener = (ScheduleNamerListener) context;
        } else {
            throw new RuntimeException(context.toString()
                    + " must implement ScheduleNamerListener");
        }
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mListener = null;
    }

    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());
        LayoutInflater inflater = getActivity().getLayoutInflater();
        View view = inflater.inflate(R.layout.dialog_add_schedule, null);
        EditTextName = view.findViewById(R.id.eText_add_schedule_name);
        String name = getArguments().getString(KEY_NAME);
        final int i = getArguments().getInt(KEY_INDEX);
        builder.setView(view)
                .setTitle(i == Action.ADD ? ADD_TITLE : RENAME_TITLE)
                .setNegativeButton("Cancel", null)
                .setPositiveButton("Ok", onOkListener(i));

        if (name != null) {
            EditTextName.setText(name);
            EditTextName.selectAll();
        }
        return builder.create();
    }

    private DialogInterface.OnClickListener onOkListener(final int i) {
        return (dialog, which) -> {
            String name = EditTextName.getText().toString();
            if (!name.isEmpty()) {
                mListener.applyName(name, i);
            }
        };
    }


    public interface ScheduleNamerListener {
        void applyName(String name, int i);
    }
}

File: app/src/main/java/com/s95ammar/weeklyschedule/views/fragments/dialogs/ScheduleNamerDialog.kt
package com.s95ammar.weeklyschedule.views.fragments.dialogs

import android.app.Dialog
import android.content.DialogInterface
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.ViewModelProviders
import com.s95ammar.weeklyschedule.R
import com.s95ammar.weeklyschedule.models.data.Schedule
import com.s95ammar.weeklyschedule.util.ListMode
import com.s95ammar.weeklyschedule.util.input
import com.s95ammar.weeklyschedule.util.toast
import com.s95ammar.weeklyschedule.viewModels.SchedulesListViewModel
import com.s95ammar.weeklyschedule.views.BlankFieldRequiredException
import com.s95ammar.weeklyschedule.views.requireNonBlankFields
import dagger.android.support.DaggerDialogFragment
import kotlinx.android.synthetic.main.dialog_edit_schedule.*
import javax.inject.Inject


class ScheduleNamerDialog : DaggerDialogFragment() {
	private var mode = ListMode.ADD
	private lateinit var editedSchedule: Schedule
	private var scheduleName = ""
		set(value) {
			field = value
			editText_edit_schedule_name.setText(scheduleName)
		}

	@Inject lateinit var factory: ViewModelProvider.Factory
	private lateinit var viewModel: SchedulesListViewModel
	private var dialogView: View? = null

	override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
		dialogView = requireActivity().layoutInflater.inflate(R.layout.dialog_edit_schedule, null)
		return AlertDialog.Builder(requireActivity())
				.setView(dialogView)
				.setTitle(R.string.schedule_add_title)
				.setNegativeButton(R.string.cancel, null)
				.setPositiveButton(R.string.ok, onOkListener())
				.create()
	}

	// Workaround for "Can't access the Fragment View's LifecycleOwner before onCreateView()"
	override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?) = dialogView

	override fun onDestroy() {
		dialogView = null
		super.onDestroy()
	}

	override fun onActivityCreated(savedInstanceState: Bundle?) {
		super.onActivityCreated(savedInstanceState)
		viewModel = ViewModelProviders.of(requireActivity(), factory).get(SchedulesListViewModel::class.java)
		initObservers()
		editText_edit_schedule_name.setText(scheduleName)
	}

	private fun initObservers() {
		viewModel.editedSchedule.observe(viewLifecycleOwner, Observer {
			it?.let { schedule ->
				editedSchedule = schedule
				setModeEdit()
			}
		})
	}

	private fun setModeEdit() {
		mode = ListMode.EDIT
		scheduleName = editedSchedule.name
		dialog?.setTitle(R.string.schedule_rename_title)
	}

	private fun onOkListener() = DialogInterface.OnClickListener { _, _ ->
		try {
			requireNonBlankFields(editText_edit_schedule_name to "schedule name")
			when (mode) {
				ListMode.ADD -> viewModel.insert(Schedule(editText_edit_schedule_name.input))
				ListMode.EDIT -> viewModel.update(getUpdatedSchedule())
			}
		} catch (e: BlankFieldRequiredException) {
			toast(e.message, Toast.LENGTH_LONG)
		}
	}

	private fun getUpdatedSchedule() = Schedule(
			editText_edit_schedule_name.input,
			editedSchedule.isActive
	).apply { id = editedSchedule.id }

	override fun onDetach() {
		viewModel.clearNamerDialogValues()
		super.onDetach()
	}
}
